### PR TITLE
Allow icmp fragmentation needed agent option

### LINKED ISSUE #8129: [NetPol] Allow ICMP messages of a certain type/code
## Proposal / RFE

As discussed with @tgraf on Slack, I'm opening an issue :

**Is your feature request related to a problem?**

Yes, Path MTU Discovery breaks when applying L4 Network Policies since ICMP unreachable messages from routers in the path are filtered.

**Describe the solution you'd like**

It would be great to be able to allow all ICMP messages of a certain type (and/or with a certain code) to be allowed. Either with a specific policy, or with an option.

For example, being able to allow ICMP/type 3/code 4 (more details [here](https://www.iana.org/assignments/icmp-parameters/icmp-parameters.xhtml#icmp-parameters-codes-3)) would prevent MTU issues.

Thanks a lot !

### PR DESCRIPTION
**Pull request by fristonio**
* Fixes #8129 
* Add a cilium-agent flag `allow-icmp-frag-needed` which allows ICMP type3 Code 4 type packets which can are used for TCP Path MTU.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8218)

### CODE REVIEW

File: daemon/daemon_main.go [hunk 1]
```diff
@@ -361,6 +361,9 @@ func init() {
 	flags.StringSlice(option.AgentLabels, []string{}, "Additional labels to identify this agent")
 	option.BindEnv(option.AgentLabels)
 
+	flags.Bool(option.AllowICMPFragNeeded, defaults.AllowICMPFragNeeded, "Annotate Kubernetes node")
```
Comment (brb) [inline daemon/daemon_main.go]: Perhaps the flag description can be a bit more descriptive?

File: bpf/lib/policy.h [hunk 2]
```diff
@@ -139,6 +139,26 @@ static inline int __inline__
...
 		    __u16 dport, __u8 proto, int dir, bool is_fragment)
 {
+#ifdef ALLOW_ICMP_FRAG_NEEDED
+	// When ALLOW_ICMP_FRAG_NEEDED is defined we allow all packets
+	// of ICMP type 3 code 4 - Fragmentation Needed
+	if (proto == IPPROTO_ICMP) {
+		void *data, *data_end;
+		struct icmphdr icmphdr;
+		struct iphdr *ip4;
+		__u32 off = ((void *)ip4 - data) + ipv4_hdrlen(ip4);
```
Comment (brb) [inline bpf/lib/policy.h]: You probably wanted to calculate the offset after calling the `revalidate_data()` below?

**Review by brb — CHANGES_REQUESTED** [review]
Thanks for the contribution! Please see my comments below.

File: bpf/lib/policy.h (hunk 2 above)
Comment (fristonio) [reply bpf/lib/policy.h]: Sorry for this.
Can you tell me like where and how can I add tests for this?

Comment (brb) [reply bpf/lib/policy.h]: You could write a smoke test which would actually do a PMTU discovery. To begin with, check https://docs.cilium.io/en/latest/contributing/#end-to-end-testing-framework and existing tests at https://github.com/cilium/cilium/tree/master/test/k8sT or https://github.com/cilium/cilium/tree/master/test/runtime.

**Comment by tgraf** [comment]
This looks fantastic

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
```
00:55:59.097      runtime: Jun 06 08:51:33 runtime cilium-agent[28160]: level=warning msg="In file included from /var/lib/cilium/bpf/bpf_overlay.c:37:" subsys=daemon
00:55:59.097      runtime: Jun 06 08:51:33 runtime cilium-agent[28160]: level=warning msg="/var/lib/cilium/bpf/lib/policy.h:147:18: error: variable has incomplete type 'struct icmphdr'" subsys=daemon
00:55:59.097      runtime: Jun 06 08:51:33 runtime cilium-agent[28160]: level=warning msg="                struct icmphdr icmphdr;" subsys=daemon
00:55:59.097      runtime: Jun 06 08:51:33 runtime cilium-agent[28160]: level=warning msg="                               ^" subsys=daemon
00:55:59.097      runtime: Jun 06 08:51:33 runtime cilium-agent[28160]: level=warning msg="/var/lib/cilium/bpf/lib/policy.h:147:10: note: forward declaration of 'struct icmphdr'" subsys=daemon
00:55:59.097      runtime: Jun 06 08:51:33 runtime cilium-agent[28160]: level=warning msg="                struct icmphdr icmphdr;" subsys=daemon
00:55:59.097      runtime: Jun 06 08:51:33 runtime cilium-agent[28160]: level=warning msg="                       ^" subsys=daemon
00:55:59.097      runtime: Jun 06 08:51:33 runtime cilium-agent[28160]: level=warning msg="/var/lib/cilium/bpf/lib/policy.h:158:22: error: use of undeclared identifier 'ICMP_DEST_UNREACH'" subsys=daemon
00:55:59.097      runtime: Jun 06 08:51:33 runtime cilium-agent[28160]: level=warning msg="                if(icmphdr.type == ICMP_DEST_UNREACH && icmphdr.code == ICMP_FRAG_NEEDED)" subsys=daemon
00:55:59.097      runtime: Jun 06 08:51:33 runtime cilium-agent[28160]: level=warning msg="                                   ^" subsys=daemon
00:55:59.097      runtime: Jun 06 08:51:33 runtime cilium-agent[28160]: level=warning msg="/var/lib/cilium/bpf/lib/policy.h:158:59: error: use of undeclared identifier 'ICMP_FRAG_NEEDED'" subsys=daemon
00:55:59.097      runtime: Jun 06 08:51:33 runtime cilium-agent[28160]: level=warning msg="                if(icmphdr.type == ICMP_DEST_UNREACH && icmphdr.code == ICMP_FRAG_NEEDED)" subsys=daemon
00:55:59.097      runtime: Jun 06 08:51:33 runtime cilium-agent[28160]: level=warning msg="                                                                        ^" subsys=daemon
00:55:59.097      runtime: Jun 06 08:51:33 runtime cilium-agent[28160]: level=warning msg="3 errors generated." subsys=daemon
```

It will make sense to add `ALLOW_ICMP_FRAG_NEEDED` to the following list in `bpf/Makefile`:

```
LXC_OPTIONS = \
         -DSKIP_DEBUG \
         -DHAVE_LPM_MAP_TYPE \
         -DHAVE_LRU_MAP_TYPE \
         -DENABLE_IPV4 \
         -DENABLE_IPV4:-DHAVE_LPM_MAP_TYPE \
         -DENABLE_IPV4:-DHAVE_LPM_MAP_TYPE:-DHAVE_LRU_MAP_TYPE \
         -DENABLE_IPV6 \
         -DENABLE_IPV6:-DHAVE_LPM_MAP_TYPE \
         -DENABLE_IPV6:-DHAVE_LPM_MAP_TYPE:-DHAVE_LRU_MAP_TYPE \
         -DENABLE_IPV6:-DENABLE_IPV4 \
         -DENABLE_IPV6:-DENABLE_IPV4:-DENABLE_ROUTING \
         -DENABLE_IPV4:-DENABLE_IPV6:-DHAVE_LPM_MAP_TYPE:-DHAVE_LRU_MAP_TYPE \
         -DENABLE_HOST_REDIRECT:-DENABLE_IPV4:-DENABLE_IPV6 \
         -DENABLE_HOST_REDIRECT:-DENABLE_IPV4:-DENABLE_IPV6:-DENABLE_NAT46
```

**Comment by tgraf** [comment]
You can then test-compile your code by running `make` in `bpf/`

**Comment by tgraf** [comment]
test-me-please

**Comment by fristonio** [comment]
Will take a look at this as soon. as I get access to my laptop.

**Comment by mrostecki** [comment]
test-me-please

**Comment by fristonio** [comment]
@tgraf I could not figure out the source of error from the CI logs. It's showing `Cancelling nested steps due to timeout` and error is in `K8s 1.10 net-next vm provisioning fail`. I am not sure if this is due to my PR or something else. Can you take a look?

**Comment by ianvernon** [comment]
@fristonio we've had some issues with timeouts affecting our CI - these are actively being worked on. I'll run the build again and see if the issue you observed reproduces. thanks for your patience!

**Comment by ianvernon** [comment]
test-me-please

**Comment by fristonio** [comment]
Thanks a lot @ianvernon 
The build still errored and was again due to a timeout.
```
06:08:02      Test NAT46 connectivity between containers [It]
06:08:02      /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/runtime-gopath/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:410
06:08:02  
06:08:02      Endpoints not ready after timeout
[2019-06-09T00:38:02.852Z]     Expected
[2019-06-09T00:38:02.852Z]         <bool>: false
[2019-06-09T00:38:02.852Z]     to be true
```
But this was while checking NAT46 connectivity, I am not sure if this is due to code changes or a timeout problem like before.
I will try to debug the issue locally once and see if I can find something.

**Comment by tgraf** [comment]
The program is now over the instruction limit:
```
Jun 09 00:33:49 runtime cilium-agent[15492]: level=error msg="Command execution failed" cmd="[tc filter replace dev lxc9e98214713e4 ingress prio 1 handle 1 bpf da obj 2074_next/bpf_lxc.o sec from-container]" error="exit status 1" subsys=datapath-loader
Jun 09 00:33:49 runtime cilium-agent[15492]: level=warning subsys=datapath-loader
Jun 09 00:33:49 runtime cilium-agent[15492]: level=warning msg="Prog section '2/10' rejected: Invalid argument (22)!" subsys=datapath-loader
Jun 09 00:33:49 runtime cilium-agent[15492]: level=warning msg=" - Type:         3" subsys=datapath-loader
Jun 09 00:33:49 runtime cilium-agent[15492]: level=warning msg=" - Attach Type:  0" subsys=datapath-loader
Jun 09 00:33:49 runtime cilium-agent[15492]: level=warning msg=" - Instructions: 4107 (11 over limit)" subsys=datapath-loader
Jun 09 00:33:49 runtime cilium-agent[15492]: level=warning msg=" - License:      GPL" subsys=datapath-loader
Jun 09 00:33:49 runtime cilium-agent[15492]: level=warning subsys=datapath-loader
Jun 09 00:33:49 runt
```

 I have a local change to split LB into a separate section to relax this. I'll push it and you can rebase on top of that.

**Comment by fristonio** [comment]
By instruction limits you mean the max instruction limits(4096) on a BPF program right?
Also, how do we split LB into separate sections, is this splitting done using tail calls?
I would like to explore a bit more regarding this, can you point me to the code sections for this?

**Review by aanm — APPROVED** [review]
Small nit, can you squash both commits? Otherwise the first commit alone does not compile. Thanks

**Review by ianvernon — APPROVED** [review]
LGTM from docs side!

**Review by joestringer — APPROVED** [review]

**Comment by tgraf** [comment]
> By instruction limits you mean the max instruction limits(4096) on a BPF program right?

Yes


> Also, how do we split LB into separate sections, is this splitting done using tail calls?

Exactly

> I would like to explore a bit more regarding this, can you point me to the code sections for this?

Sure, I've pushed the branch here:
https://github.com/cilium/cilium/compare/pr/tgraf/split-lxc-prog

**Comment by borkmann** [comment]
test-me-please

**Comment by brb** [comment]
CI failed due to the same issue with instr count:

```
Jun 12 15:45:35 runtime cilium-agent[17494]: level=warning msg="Prog section '2/10' rejected: Invalid argument (22)!" subsys=datapath-loader
Jun 12 15:45:35 runtime cilium-agent[17494]: level=warning msg=" - Type:         3" subsys=datapath-loader
Jun 12 15:45:35 runtime cilium-agent[17494]: level=warning msg=" - Attach Type:  0" subsys=datapath-loader
Jun 12 15:45:35 runtime cilium-agent[17494]: level=warning msg=" - Instructions: 4107 (11 over limit)" subsys=datapath-loader
Jun 12 15:45:35 runtime cilium-agent[17494]: level=warning msg=" - License:      GPL" subsys=datapath-loader
```

**Comment by joestringer** [comment]
@brb yeah, the LXC prog splitout will need to be done first. @fristonio , will you look into that? Happy to provide pointers/help if you have any questions.

**Comment by fristonio** [comment]
Yeah sure, I would love to. I am not very experienced with BPF but will surely try to help out.

**Comment by aanm** [comment]
@fristonio needs rebase against master.

**Comment by mrostecki** [comment]
test-me-please

**Review by vadorovsky — APPROVED** [review]

**Comment by aanm** [comment]
@fristonio needs a new rebase. It seems the tests failed last time. Can you rebase so we can trigger the CI again?

**Comment by brb** [comment]
@fristonio Let us know if we should do the rebase.

**Comment by fristonio** [comment]
Sorry, I was a bit busy the past week.
Also, I think since [this](https://github.com/cilium/cilium/compare/pr/tgraf/split-lxc-prog) change is not in the master yet the tests will fail again.

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
@fristonio This looks ready to merge, just needs a rebase.

**Comment by fristonio** [comment]
@tgraf I have rebased the pull request, but I am not sure if this will pass the ginkgo tests as I think that the split-lxc-prog branch changes are still not in master.

**Comment by tgraf** [comment]
test-me-please

**Pull request merged**